File: Code/Components/DanpheEMR.ServerModel/PharmacyModels/PHRMInvoiceReturnModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel.PharmacyModels
{
    public class PHRMInvoiceReturnModel
    {
        [Key]
        public int InvoiceReturnId { get; set; }
        public int? InvoiceId { get; set; }
        public int StoreId { get; set; }
        public int? PatientId { get; set; }
        public int? CounterId { get; set; }
        public int? CreditNoteId { get; set; }
        public decimal? SubTotal { get; set; }
        public decimal? DiscountAmount { get; set; }
        public decimal? VATAmount { get; set; }
        public decimal? TotalAmount { get; set; }
        public decimal? PaidAmount { get; set; }
        public decimal? Tender { get; set; }
        public decimal? Change { get; set; }
        public int? PrintCount { get; set; }
        public decimal? Adjustment { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public bool? IsRealtime { get; set; }
        public bool? IsRemoteSynced { get; set; }
        public bool? IsTransferredToACC { get; set; }
        public virtual List<PHRMInvoiceReturnItemsModel> InvoiceReturnItems { get; set; }
        public string PaymentMode { get; set; }
        public int? FiscalYearId { get; set; }
        public string Remarks { get; set; }

        //sud:1-Oct'21--Changing Claimcode from String to Int64-- to use Incremental logic (max+1)
        //need nullable since ClaimCode is Non-Mandatory for normal visits.
        public Int64? ClaimCode { get; set; }
        public bool? IsManualReturn { get; set; }
        public string ReferenceInvoiceNo { get; set; }
        public DateTime? ReferenceInvoiceDate { get; set; }
        [NotMapped]
        public decimal? VATPercentage { get; set; }
        [NotMapped]
        public decimal? TaxableAmount { get; set; }
        [NotMapped]
        public decimal? NonTaxableAmount { get; set; }

        public int? SettlementId { get; set; }

        [NotMapped]
        public double? CashDiscount { get; set; }

    }
}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/pharmacy/shared/phrm-invoice-return.model .ts
import {
    NgForm,
    FormGroup,
    FormControl,
    Validators,
    FormBuilder,
    ReactiveFormsModule
} from '@angular/forms'
import * as moment from 'moment/moment';

import { PHRMInvoiceReturnItemsModel } from './phrm-invoice-return-items.model'
import { PHRMNarcoticRecordModel } from './phrm-narcotic-record';
import { PHRMPatient } from "./phrm-patient.model";
export class PHRMInvoiceReturnModel {
    public InvoiceReturnId: number = 0;
    public InvoiceId: number = 0;
    public StoreId: number;
    public PatientId: number = null;
    public CounterId: number = null;
    public CreditNoteId: number = 0;
    public SubTotal: number = 0;
    public Tender: number = 0;
    public Change: number = 0;
    public DiscountAmount: number = 0;
    public DiscountPercentage: number = 0;
    public VATAmount: number = 0;
    public TotalAmount: number = 0;
    public PaidAmount: number = 0;
    public CreatedBy: number = 0;
    public CreatedOn: string = "";
    public Adjustment: number = 0;
    public PrintCount: number = 0;
    public InvoiceReturnItems: Array<PHRMInvoiceReturnItemsModel> = new Array<PHRMInvoiceReturnItemsModel>();
    public DepositDeductAmount: number = 0;
    public DepositAmount: number = 0;
    public DepositBalance: number = 0;
    public PaymentMode: string = "";
    public Remarks: string = "";
    //public InvoiceCode: string = "PHRM";
    public FiscalYearId: number;
    //only for read purpose
    public selectedPatient: PHRMPatient = new PHRMPatient();

    //only for show in list
    public PatientName: string = "";

    public InvoiceReturnValidator: FormGroup = null;
    public ClaimCode: number;
    // added for Manual Return
    IsManualReturn: boolean;
    ReferenceInvoiceNo: string;
    ReferenceInvoiceDate: string = moment().format('YYYY-MM-DD');
    VATPercentage: number = 0;
    TaxableAmount: number = 0;
    NonTaxableAmount: number = 0;

    CashDiscount: number = 0;
    SettlementId: number = null;
    //Constructor of class
    constructor() {
        var _formBuilder = new FormBuilder();
        this.InvoiceReturnValidator = _formBuilder.group({
            'Remark': ['', Validators.compose([Validators.required])]
        });
    }
    public IsDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.InvoiceReturnValidator.dirty;
        else
            return this.InvoiceReturnValidator.controls[fieldName].dirty;
    }


    public IsValid(): boolean {
        if (this.InvoiceReturnValidator.valid) {
            return true;
        } else {
            return false;
        }
    }

    public IsValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined)
            return this.InvoiceReturnValidator.valid;
        else
            return !(this.InvoiceReturnValidator.hasError(validator, fieldName));
    }

    public EnableControl(formControlName: string, enabled: boolean) {
        let currCtrol = this.InvoiceReturnValidator.controls[formControlName];
        if (currCtrol) {
            if (enabled) {
                currCtrol.enable();
            }
            else {
                currCtrol.disable();
            }
        }
    }
}
